src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs

```diff
@@ -0,0 +1,294 @@
+        {
+            IReliableConcurrentQueue<ReleasePipelineRunnerItem> queue =
+                await StateManager.GetOrAddAsync<IReliableConcurrentQueue<ReleasePipelineRunnerItem>>("queue");
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    using (ITransaction tx = StateManager.CreateTransaction())
+                    {
+                        ConditionalValue<ReleasePipelineRunnerItem> maybeItem = await queue.TryDequeueAsync(
+                            tx,
+                            cancellationToken);
+                        if (maybeItem.HasValue)
+                        {
+                            ReleasePipelineRunnerItem item = maybeItem.Value;
+                            using (Logger.BeginScope(
+                                "Triggering release pipelines associated with channel {channelId} for build {buildId}.",
+                                item.BuildId,
+                                item.ChannelId))
+                            {
+                                await RunAssociatedReleasePipelinesAsync(item.BuildId, item.ChannelId, cancellationToken);
+                            }
+                        }
+                        else
+                        {
+                            await Task.Delay(1000, cancellationToken);
+                        }
+                        await tx.CommitAsync();
+                    }
+                }
+                catch (TaskCanceledException tcex) when (tcex.CancellationToken == cancellationToken)
+                {
+                    // ignore
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogError(ex, "Processing queue messages");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Start release pipeline associated with a channel.
+        /// </summary>
+        /// <param name="buildId">Maestro build id.</param>
+        /// <param name="channelId">Maestro channel id.</param>
+        /// <returns></returns>
+        public async Task RunAssociatedReleasePipelinesAsync(int buildId, int channelId, CancellationToken cancellationToken)
+        {
+            Logger.LogInformation($"Starting release pipeline for {buildId} in {channelId}");
+            Build build = await Context.Builds
+                .Where(b => b.Id == buildId).FirstOrDefaultAsync();
+
+            Channel channel = await Context.Channels
+                .Where(ch => ch.Id == channelId)
+                .Include(ch => ch.ChannelReleasePipelines)
+                .ThenInclude(crp => crp.ReleasePipeline)
+                .FirstOrDefaultAsync();
+
+            if (build == null || channel == null)
+            {
+                Logger.LogInformation($"Could not find the specified build {buildId} or channel {channelId} to run a release pipeline on");
```

Comment:
Should this be an error? At least if build == null, since builds can't be deleted.  Channels can be deleted, but it's a pretty exceptional case.

Reply:
I'll split it and make it an error if Build is null.